ipam/allocator.go

```diff
@@ -0,0 +1,610 @@
+			// Release it
+			for {
+				var err error
+				if err = space.PushReservation(ordinal/8, ordinal%8, true); err == nil {
+					break
+				}
+				if _, ok := err.(types.RetryError); ok {
+					// bitmask must have changed, retry delete
+					continue
+				}
+				log.Warnf("Failed to release address %s because of internal error: %s", address.String(), err.Error())
+				return
+			}
+			return
+		}
+
+	}
+}
+
+func (a *Allocator) reserveAddress(addrSpace AddressSpace, subnet *net.IPNet, prefAddress net.IP, ver ipVersion) (net.IP, *net.IPNet, error) {
+	var keyList []subnetKey
+
+	// Get the list of pointers to the internal subnets
+	if subnet != nil {
+		// Get the list of smaller internal subnets
+		subnetList, err := getInternalSubnets(subnet, a.internalHostSize)
+		if err != nil {
+			return nil, nil, err
+		}
+		for _, s := range subnetList {
+			keyList = append(keyList, subnetKey{addrSpace, subnet.String(), s.String()})
+		}
+	} else {
+		a.Lock()
+		keyList = a.getSubnetList(addrSpace, ver)
+		a.Unlock()
+	}
+	if len(keyList) == 0 {
+		return nil, nil, ErrNoAvailableSubnet
+	}
+
+	for _, key := range keyList {
+		a.Lock()
+		bitmask := a.addresses[key]
+		a.Unlock()
+		address, err := a.getAddress(key.canonicalChildSubnet(), bitmask, prefAddress, ver)
+		if err == nil {
+			return address, subnet, nil
+		}
+	}
+
+	return nil, nil, ErrNoAvailableIPs
+}
+
+// Get the list of available internal subnets for the specified address space and the desired ip version
+func (a *Allocator) getSubnetList(addrSpace AddressSpace, ver ipVersion) []subnetKey {
+	var list [1024]subnetKey
+	ind := 0
+	a.Lock()
+	for subKey := range a.addresses {
+		s := subKey.canonicalSubnet()
+		subVer := getAddressVersion(s.IP)
+		if subKey.addressSpace == addrSpace && subVer == ver {
+			list[ind] = subKey
+			ind++
+		}
+	}
+	a.Unlock()
+	return list[0:ind]
+}
+
+func (a *Allocator) getAddress(subnet *net.IPNet, bitmask *bitseq.Handle, prefAddress net.IP, ver ipVersion) (net.IP, error) {
+	var (
+		bytePos, bitPos int
+		ordinal         int
+		err             error
+	)
+	// Look for free IP, skip .0 and .255, they will be automatically reserved
+again:
+	if bitmask.Unselected() <= 0 {
+		return nil, ErrNoAvailableIPs
+	}
+	if prefAddress == nil {
+		bytePos, bitPos, err = bitmask.GetFirstAvailable()
+	} else {
+		ordinal = ipToInt(getHostPortionIP(prefAddress, subnet))
+		bytePos, bitPos, err = bitmask.CheckIfAvailable(ordinal)
+	}
+	if err != nil {
+		return nil, ErrNoAvailableIPs
+	}
+
+pushsame:
+	// Lock it
+	if err = bitmask.PushReservation(bytePos, bitPos, false); err != nil {
+		if _, ok := err.(types.RetryError); !ok {
+			return nil, fmt.Errorf("internal failure while reserving the address: %s", err.Error())
+		}
+		// bitmask view must have changed. Selected address may or may no longer be available
+		if prefAddress != nil {
+			if _, _, err = bitmask.CheckIfAvailable(ordinal); err == nil {
+				//still available
+				goto pushsame
+			}
+			goto again
```

Comment:
Can we please refactor this code to not use goto this way of jumping back? Instead a for loop with `continue` and `break